File: MinimumPenaltyForAShop/minimum_penalty_for_a_shop.dart
/*


-* 2483. Minimum Penalty for a Shop *-

You are given the customer visit log of a shop represented by a 0-indexed string customers consisting only of characters 'N' and 'Y':

if the ith character is 'Y', it means that customers come at the ith hour
whereas 'N' indicates that no customers come at the ith hour.
If the shop closes at the jth hour (0 <= j <= n), the penalty is calculated as follows:

For every hour when the shop is open and no customers come, the penalty increases by 1.
For every hour when the shop is closed and customers come, the penalty increases by 1.
Return the earliest hour at which the shop must be closed to incur a minimum penalty.

Note that if a shop closes at the jth hour, it means the shop is closed at the hour j.

 

Example 1:

Input: customers = "YYNY"
Output: 2
Explanation: 
- Closing the shop at the 0th hour incurs in 1+1+0+1 = 3 penalty.
- Closing the shop at the 1st hour incurs in 0+1+0+1 = 2 penalty.
- Closing the shop at the 2nd hour incurs in 0+0+0+1 = 1 penalty.
- Closing the shop at the 3rd hour incurs in 0+0+1+1 = 2 penalty.
- Closing the shop at the 4th hour incurs in 0+0+1+0 = 1 penalty.
Closing the shop at 2nd or 4th hour gives a minimum penalty. Since 2 is earlier, the optimal closing time is 2.
Example 2:

Input: customers = "NNNNN"
Output: 0
Explanation: It is best to close the shop at the 0th hour as no customers arrive.
Example 3:

Input: customers = "YYYY"
Output: 4
Explanation: It is best to close the shop at the 4th hour as customers arrive at each hour.
 

Constraints:

1 <= customers.length <= 105
customers consists only of characters 'Y' and 'N'.

*/

class Solution {
  int bestClosingTime(String customers) {
    int yes = 0;
    final int n = customers.length;
    for (int i = 0; i < n; i++) {
      if (customers[i] == 'Y') yes++;
    }
    int ans = 0;
    int pen = yes;
    int no = 0;
    for (int i = 0; i < n; i++) {
      if (customers[i] == 'N') {
        no++;
      } else
        yes--;
      int currentPen = yes + no;
      if (currentPen < pen) {
        ans = i + 1;
        pen = currentPen;
      }
    }
    return ans;
  }
}

File: MinimumPenaltyForAShop/minimum_penalty_for_a_shop.go
package main

/*
 int yes = 0;
    final int n = customers.length;
    for (int i = 0; i < n; i++) {
      if (customers[i] == 'Y') yes++;
    }
    int ans = 0;
    int pen = yes;
    int no = 0;
    for (int i = 0; i < n; i++) {
      if (customers[i] == 'N') {
        no++;
      } else
        yes--;
      int currentPen = yes + no;
      if (currentPen < pen) {
        ans = i + 1;
        pen = currentPen;
      }
    }
    return ans;


*/

func bestClosingTime(customers string) int {
	var yes int = 0
	var n int = len(customers)
	for i := 0; i < n; i++ {
		if customers[i] == 'Y' {
			yes++
		}
	}
	var ans int = 0
	var pen int = yes
	var no int = 0
	for i := 0; i < n; i++ {
		if customers[i] == 'N' {
			no++
		} else {
			yes--
		}
		var currentPen int = yes + no
		if currentPen < pen {
			ans = i + 1
			pen = currentPen
		}
	}

	return ans
}
